File: subwrap.pl
#!/usr/bin/perl

sub playfile {
	my($file, $sub, $encoding, $args) = @_;
	if ($file =~ m/.*\.wmv/) {
		$args .= " -subdelay -5 "
	}
	if ($encoding eq "")
	{
		system("mplayer \"$file\" -sub \"$sub\" -utf8 $args");
	} else {
#		system("iconv -f $encoding -c \"$sub\" | mplayer \"$file\" -sub - -utf8 $args");
		system("iconv -f $encoding -c \"$sub\" > /tmp/subtitle.sub");
		system("mplayer \"$file\" -ass -sub /tmp/subtitle.sub -utf8 $args");
		system("rm -f /tmp/subtitle.sub");
	}
}

@exts = ('srt','ssa','ass', '[VeryCD.com].ssa');

@langs = ('en','Eng','chs','Chs','cn','sc_v2', 'sc','SC','uni_gb','gb','GBK','GB','cht','tc','TC','uni_big5','neta');

if ($#ARGV == -1)
{
	print "Usage: subwrap <filename>\n";
	exit 1;
}

$fname = $ARGV[0];

$argstr = "";

for ($i = 1; $i <= $#ARGV; $i = $i + 1)
{
	$argstr = $argstr . $ARGV[$i] . " ";
}

if (!(-e $fname))
{
	print "Cannot find specified file.\n";
	exit 2;
}

if ($ENV{'SUBFILE'})
{
	$subname = $ENV{'SUBFILE'};
	goto FOUND;
}

if (!($fname =~ m/(.*)\.(.*)/))
{
	print "Cannot find . in filename.\n";
	exit 1;
}


for ($way = 0; $way < 3; $way++)
{
	if ($way == 0)
	{
		$fname =~ m/(.*)\.(.*)/;
		$basename = $1;
	} elsif ($way == 1) {
		$fname =~ m/(.*)\[.*\]\.(.*)/;
		$basename = $1;
	} elsif ($way == 2) {
		$fname =~ m/(.*)\.\[.*\]\.(.*)/;
		$basename = $1;
	}

	# First let's try only extensions
	foreach $i (@exts)
	{
		$subname = $basename . "." . $i;
		if (-e $subname)
		{
			goto FOUND;
		}
	}

	# Now let's try extensions with languages
	foreach $i (@langs)
	{
		foreach $j (@exts)
		{
			$subname = $basename . "." . $i . "." . $j;
			if (-e $subname)
			{
				goto FOUND;
			}
		}
	}
}
	
# We failed!
print "Cannot found subtitles in (@exts) or (@langs) . (@exts).\n";
exit 2;

FOUND:

$filetype = `file "$subname"`;
print "Found sub file: $subname\n";

if ($filetype =~ m/ISO-8859/ or $filetype =~ m/extended-ASCII/)
{
	# GBK file
	print "Converting from GBK encoding\n";
	playfile($fname, $subname, "gbk", $argstr);
} elsif ($filetype =~ m/UTF-16/) {
	# UTF-16 file
	if ($filetype =~ m/Little-endian/)
	{
		# UTF-16 Little Endian
		print "Converting from UTF-16LE encoding\n";
		playfile($fname, $subname, "utf-16le", $argstr);
	} else {
		# UTF-16 Big Endian
		print "Converting from UTF-16BE encoding\n";
		playfile($fname, $subname, "utf-16be", $argstr);
	}
} else {
	playfile($fname, $subname, "", $argstr);
}

File: subwrap.py
#!/usr/bin/env python3

import sys
import logging
import os
import glob
import collections
import subprocess

logger = logging.getLogger("subwrap")

EXTS = ('srt','ssa','ass')
LANGS = {
    'en': ('en','eng'),
    'chs': ('chs','cn','sc_v2', 'sc','SC','uni_gb','gb','gbk'),
    'cht': ('cht','chi','tc','uni_big5'),
}

def play(path):
    logger.info("playing %s", path)
    path = os.path.abspath(path)
    logger.debug("Full path: %s", path)

    dirname = os.path.dirname(path)
    basename = os.path.basename(path)
    logger.debug("Dir: %s", dirname)
    logger.debug("Basename: %s", basename)

    (name, ext) = os.path.splitext(basename)
    logger.debug("Name: %s", name)
    logger.debug("Extension: %s", ext)

    subtitles = collections.defaultdict(list)
    for otherfile in os.listdir(dirname):
        if otherfile == basename: continue
        if otherfile.startswith(name):
            (othername, otherext) = os.path.splitext(otherfile)
            otherext = otherext.lower()
            if otherext.startswith('.'): otherext = otherext[1:]
            if otherext in EXTS:
                (othername2, otherext2) = os.path.splitext(othername)
                otherext2 = otherext2.lower()
                if otherext2.startswith('.'): otherext2 = otherext2[1:]
                logger.debug("other: %s %s", othername2, otherext2)

                sublang = None
                for lang, exts in LANGS.items():
                    if otherext2 in exts:
                        sublang = lang
                        break
                if sublang is None:
                    for lang, exts in LANGS.items():
                        for ext in exts:
                            if otherext2.startswith(ext):
                                sublang = lang
                                break
                        if sublang is not None: break
                if sublang is None: sublang = 'unknown'
                logger.debug("Language: %s", sublang)

                subtitles[sublang].append(otherfile)

    for sublang, subfiles in subtitles.items():
        for subfile in subfiles:
            logger.info("lang %s: %s", sublang, subfile)

    chosen = None
    chosen_lang = 'unknown'
    for lang in LANGS:
        if lang in subtitles:
            chosen = subtitles[lang][0]
            chosen_lang = lang
            break

    if chosen is None:
        if subtitles:
            chosen = list(subtitles.values())[0][0]

    if chosen is None:
        logger.error("Fail to found a subtitle for %s", path)
        subprocess.check_call(["mpv", path])
    else:
        logger.info("Chosen %s, lang %s", chosen, chosen_lang)
        chosen = os.path.join(dirname, chosen)
        subprocess.check_call(["mpv", "--sub-file=%s" % chosen, path])


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    if len(sys.argv) < 2:
        logger.error("Usage: %s <media file>", sys.argv[0])
        sys.exit(1)

    path = sys.argv[1]
    play(path)
